doc/developer/design/20210401_volatile_sources.md

```diff
@@ -0,0 +1,337 @@
+--------------+------+----------+-------------
+ pubnub_view  | user | true     | true
+ kafka_view   | user | false    | true
+```
+
+In the docs, we'll want to add a page about "Volatile sources" that explains
+what they are, why they are good for demos, and why they are bad for production
+applications that care about correctness. We can add a warning callout on
+PubNub, SSE, and Kinesis source pages that links to the volatile source page.
+
+The big "gotcha" with volatile sources is that different instantiations can have
+different data within the same `materialized` process. Consider a sequence like
+the following:
+
+```sql
+CREATE SOURCE market_orders_raw FROM PUBNUB
+SUBSCRIBE KEY 'sub-c-4377ab04-f100-11e3-bffd-02ee2ddab7fe'
+CHANNEL 'pubnub-market-orders';
+
+CREATE MATERIALIZED VIEW market_orders_1 AS
+  SELECT text::jsonb AS val FROM market_orders_raw;
+
+-- Imagine the user makes a cup of coffee now.
+
+CREATE MATERIALIZED VIEW market_orders_2 AS
+  SELECT text::jsonb AS val FROM market_orders_raw;
+```
+
+The data in `market_orders_1` and `market_orders_2` will not be equivalent,
+because they'll each have instantiated separate copies of the source at
+separate times. `market_orders_1` will contain some historical data that
+`market_orders_2` does not.
+
+The more I think about it, the more this seems like reasonable, even desirable
+behavior from a volatile source, provided we can educate users sufficiently.
+
+### Implications
+
+#### Exactly-once sinks
+
+Exactly-once sinks will need to error if they are created on a volatile source
+or view. This will be easy to check in the coordinator. The error message might
+read:
+
+```
+ERROR:  exactly-once sinks cannot be created from a volatile {source|view}
+DETAIL: Data in volatile {sources|views} is not consistent across restarts,
+        which is at odds with exactly-once semantics. To learn more, visit
+        https://materialize.com/s/volatility.
+```
+
+#### Active-active replication
+
+Active-active replication will rely on the determinism of the dataflow
+computation to compute identical results across two separate instances of
+Materialize. That is at odds with volatile sources, which will provide the
+two copies of Materialize with two different sets of data.
+
+One easy solution is to simply ban volatile sources when using Materialize
+in this hypothetical distributed mode.
+
+Alternatively, we could say "you got what you asked for", and message the
+deficiencies accordingly.
+
+#### Optimizer
+
+The optimizer will need to be careful to treat different instances of volatile
+sources and views as logically different views. This is a burden for the
+optimizer team, but hopefully a small one. The check to see whether two
+sources or views is roughly:
+
+```rust
+fn are_objects_equivalent(metadata: &Metadata, id1: GlobalId, id2: GlobalId) -> bool {
+    id1 == id2 && !metadata.is_volatile(id1)
+}
+```
+
+## Future work
+
+### Toggleable volatility
```

Comment:
What does this toggle imply? That users accept a certain level of risk with `volatile` sources or that they want to be prevented from doing unreasonable things with `nonvolatile` ones?